### PR TITLE
Apply ad-hoc filters to PPL queries before sending it to the backend

### LINKED ISSUE #226: Frontend: Pre-requisite for PPL migrations - refactor adhoc filters on frontend before sending query to backend
Similar to https://github.com/grafana/opensearch-datasource/pull/225, but for PPL

### PR DESCRIPTION
**Pull request by kevinwcyu**
**What this PR does / why we need it**:

Prerequisite work for migrating PPL queries to be executed on the backend. Ad-hoc filters are applied on the frontend only so this work ensures they're added before sending a query to the backend.

**Which issue(s) this PR fixes**:

Fixes #226 

**Special notes for your reviewer**:

You currently can't test this except for running the unit tests since no PPL queries have been migrated to execute on the backend yet.

### CODE REVIEW

File: src/datasource.test.ts [hunk 1]
```diff
@@ -1584,60 +1584,123 @@ describe('OpenSearchDatasource', function(this: any) {
...
-        expect(query).toBe('bar:"baz" AND -job:"grafana" AND service:/service/ AND count:>1');
+      describe('PPL queries', () => {
+        const adHocFilters = [
+          { key: 'bar', operator: '=', value: 'baz', condition: '' },
+          { key: 'job', operator: '!=', value: 'grafana', condition: '' },
+          { key: 'bytes', operator: '>', value: 50, condition: '' },
+          { key: 'count', operator: '<', value: 100, condition: '' },
+          { key: 'timestamp', operator: '=', value: '2020-11-22 16:40:43', condition: '' },
+        ];
+        it('should return query with ad-hoc filters applied', () => {
```
Comment (iwysiu) [inline src/datasource.test.ts:1695]: Should this also have the query is empty case or is that not possible for ppl queries?

**Review by iwysiu — APPROVED** [review]
Approved with a question about ppl queries

File: src/datasource.test.ts [hunk 2]
```diff
@@ -1584,60 +1584,123 @@ describe('OpenSearchDatasource', function(this: any) {
...
+          const query = ctx.ds.addAdHocFilters({ refId: 'A', query: 'source = test-index', queryType: QueryType.PPL }, [
+            { key: 'a', operator: '=~', value: '', condition: '' },
+            { key: 'a', operator: '!~', value: '', condition: '' },
+          ]);
+          expect(query).toBe('source = test-index');
+        });
       });
     });
 
     describe('with 1 ad hoc filter', () => {
```
Comment (katebrenner) [inline src/datasource.test.ts:1637]: just curious - any reason why only testing Lucene in this test case?

**Review by katebrenner — APPROVED** [review]

File: src/datasource.test.ts (hunk 2 above)
Comment (kevinwcyu) [reply src/datasource.test.ts:1637]: i didn't think there was anything that could be added under this test case with 1 ad hoc filter for PPL queries that couldn't be covered by the multiple ad hoc filters test cases.

File: src/datasource.test.ts (hunk 1 above)
Comment (kevinwcyu) [reply src/datasource.test.ts:1695]: right now the flow of the PPL execution is to always make sure the query string is not empty by adding the index that is being searched to the query string if the query string is empty

https://github.com/grafana/opensearch-datasource/blob/6e6f9affd5bfbf66ac939a90768782dc7b176ce3/src/datasource.ts#L741-L743

this happens happens before ad-hoc filters are added to the query string, so for this PR, i'm just testing the adding of the ad-hoc filters.

an empty query string test case will probably be more suited for https://github.com/grafana/opensearch-datasource/issues/200

in short, ppl queries should not be empty, but that should be handled outside of the `addAdHocFilters` code.

File: src/datasource.ts [hunk 3]
```diff
@@ -396,6 +404,31 @@ export class OpenSearchDatasource extends DataSourceWithBackend<OpenSearchQuery,
     return [query, ...osFilters].filter(f => f).join(' AND ');
   }
 
+  addPPLAdHocFilters(queryString: string, adHocFilters: any) {
+    for (let i = 0; i < adHocFilters.length; i++) {
+      const { key, operator } = adHocFilters[i];
+      let { value } = adHocFilters[i];
+
+      if ('=~' === operator || '!~' === operator || !key || !value) {
```
Comment (fridgepoet) [inline src/datasource.ts:412]: Would you explain more about this condition? What's that about the operators? Was this already illustrated in a test case and I missed it?

File: src/datasource.ts [hunk 4]
```diff
@@ -396,6 +404,31 @@ export class OpenSearchDatasource extends DataSourceWithBackend<OpenSearchQuery,
     return [query, ...osFilters].filter(f => f).join(' AND ');
   }
 
+  addPPLAdHocFilters(queryString: string, adHocFilters: any) {
+    for (let i = 0; i < adHocFilters.length; i++) {
+      const { key, operator } = adHocFilters[i];
```
Comment (fridgepoet) [inline src/datasource.ts:409]: Just curious: I notice in the existing `addPPLAdhocFilters` in `buildPPLQuery` that they're always using `adhocFilters[i].value`. How come we don't have to use `.value` here?

**Review by fridgepoet — APPROVED** [review]
Nice!

File: src/datasource.ts (hunk 3 above)
Comment (kevinwcyu) [reply src/datasource.ts:412]: this wasn't in a test case before. these two operators are not recognized by opensearch so it returns an error if you try to use them.

File: src/datasource.ts (hunk 4 above)
Comment (kevinwcyu) [reply src/datasource.ts:409]: using `let { value } = adHocFilters[i];` is shorter and easier to read in the later code since you can just use `value` instead of `adhocFilters[i].value`.

File: src/datasource.ts (hunk 3 above)
Comment (fridgepoet) [reply src/datasource.ts:412]: are there tests now documenting this? if not, would it be worth it to add them? honestly asking that and deferring to your judgment about this!

Comment (kevinwcyu) [reply src/datasource.ts:412]: https://github.com/grafana/opensearch-datasource/pull/244/files#diff-f5855f08029d58a7871f69a0ebfe07376b2508725c81159df890fc793eb5b545R1629-R1630 there's this test case
